{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyP5AUJUhqQpcdYF+xyYz+eC",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU",
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/xuebingwu/splice-site-prediction-with-pangolin/blob/main/predict_splice_sites_in_a_sequence.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Predict 5' and 3' splice sites from a single sequence using *Pangolin*\n",
        "* Given a DNA/RNA sequence, predict potential 5' and 3' splice sites and their tissue-specific usage \n",
        "* Use deep learning models trained in four human tissues: heart, liver, brain, and testis (*Pangolin*: [Zeng and Li, Genome Biology 2022](https://doi.org/10.1186/s13059-022-02664-4)) \n",
        "* To run: enter your sequence, then click `Runtime` -> `Run all` in the menu bar\n",
        "* Speed/time: the first run takes ~50s (mostly installing the package). Subsequent runs ~2s\n"
      ],
      "metadata": {
        "id": "HBC_J9pHSp1p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Analyze your sequence\n",
        "\n",
        "# scripts from custom_usage.py from the Pangolin package\n",
        "\n",
        "import os\n",
        "from matplotlib import pylab\n",
        "import matplotlib.lines as mlines\n",
        "from google.colab import files\n",
        "import datetime\n",
        "\n",
        "\n",
        "########## input \n",
        "SEQUENCE = \"Enter your sequence or select a sample sequence (default: ACTB pre-mRNA)\" #@param [\"ACTB pre-mRNA\", \"GAPDH pre-mRNA\", \"EGFP + CENPB 3' UTR\"] {allow-input: true}\n",
        "\n",
        "#@markdown - To run: click `Runtime` -> `Run all` in the menu bar, or click the triangle play/run button on the left\n",
        "#@markdown - Input format: raw sequence (A/C/G/T/U only); can paste a multi-line sequence from a FASTA file (without the header) \n",
        "#@markdown - More accurate to use full-length transcript sequence (model scores each 10kb window)\n",
        "#@markdown - Can also run on a short sequence \n",
        "\n",
        "seq = SEQUENCE\n",
        "\n",
        "if seq == \"ACTB pre-mRNA\":\n",
        "  seq = \"CGAGACCGCGTCCGCCCCGCGAGCACAGAGCCTCGCCTTTGCCGATCCGCCGCCCGTCCACACCCGCCGCCAGGTAAGCCCGGCCAGCCGACCGGGGCAGGCGGCTCACGGCCCGGCCGCAGGCGGCCGCGGCCCCTTCGCCCGTGCAGAGCCGCCGTCTGGGCCGCAGCGGGGGGCGCATGGGGGGGGAACCGGACCGCCGTGGGGGGCGCGGGAGAAGCCCCTGGGCCTCCGGAGATGGGGGACACCCCACGCCAGTTCGGAGGCGCGAGGCCGCGCTCGGGAGGCGCGCTCCGGGGGTGCCGCTCTCGGGGCGGGGGCAACCGGCGGGGTCTTTGTCTGAGCCGGGCTCTTGCCAATGGGGATCGCAGGGTGGGCGCGGCGGAGCCCCCGCCAGGCCCGGTGGGGGCTGGGGCGCCATTGCGCGTGCGCGCTGGTCCTTTGGGCGCTAACTGCGTGCGCGCTGGGAATTGGCGCTAATTGCGCGTGCGCGCTGGGACTCAAGGCGCTAACTGCGCGTGCGTTCTGGGGCCCGGGGTGCCGCGGCCTGGGCTGGGGCGAAGGCGGGCTCGGCCGGAAGGGGTGGGGTCGCCGCGGCTCCCGGGCGCTTGCGCGCACTTCCTGCCCGAGCCGCTGGCCGCCCGAGGGTGTGGCCGCTGCGTGCGCGCGCGCCGACCCGGCGCTGTTTGAACCGGGCGGAGGCGGGGCTGGCGCCCGGTTGGGAGGGGGTTGGGGCCTGGCTTCCTGCCGCGCGCCGCGGGGACGCCTCCGACCAGTGTTTGCCTTTTATGGTAATAACGCGGCCGGCCCGGCTTCCTTTGTCCCCAATCTGGGCGCGCGCCGGCGCCCCCTGGCGGCCTAAGGACTCGGCGCGCCGGAAGTGGCCAGGGCGGGGGCGACCTCGGCTCACAGCGCGCCCGGCTATTCTCGCAGCTCACCATGGATGATGATATCGCCGCGCTCGTCGTCGACAACGGCTCCGGCATGTGCAAGGCCGGCTTCGCGGGCGACGATGCCCCCCGGGCCGTCTTCCCCTCCATCGTGGGGCGCCCCAGGCACCAGGTAGGGGAGCTGGCTGGGTGGGGCAGCCCCGGGAGCGGGCGGGAGGCAAGGGCGCTTTCTCTGCACAGGAGCCTCCCGGTTTCCGGGGTGGGGGCTGCGCCCGTGCTCAGGGCTTCTTGTCCTTTCCTTCCCAGGGCGTGATGGTGGGCATGGGTCAGAAGGATTCCTATGTGGGCGACGAGGCCCAGAGCAAGAGAGGCATCCTCACCCTGAAGTACCCCATCGAGCACGGCATCGTCACCAACTGGGACGACATGGAGAAAATCTGGCACCACACCTTCTACAATGAGCTGCGTGTGGCTCCCGAGGAGCACCCCGTGCTGCTGACCGAGGCCCCCCTGAACCCCAAGGCCAACCGCGAGAAGATGACCCAGGTGAGTGGCCCGCTACCTCTTCTGGTGGCCGCCTCCCTCCTTCCTGGCCTCCCGGAGCTGCGCCCTTTCTCACTGGTTCTCTCTTCTGCCGTTTTCCGTAGGACTCTCTTCTCTGACCTGAGTCTCCTTTGGAACTCTGCAGGTTCTATTTGCTTTTTCCCAGATGAGCTCTTTTTCTGGTGTTTGTCTCTCTGACTAGGTGTCTAAGACAGTGTTGTGGGTGTAGGTACTAACACTGGCTCGTGTGACAAGGCCATGAGGCTGGTGTAAAGCGGCCTTGGAGTGTGTATTAAGTAGGTGCACAGTAGGTCTGAACAGACTCCCCATCCCAAGACCCCAGCACACTTAGCCGTGTTCTTTGCACTTTCTGCATGTCCCCCGTCTGGCCTGGCTGTCCCCAGTGGCTTCCCCAGTGTGACATGGTGTATCTCTGCCTTACAGATCATGTTTGAGACCTTCAACACCCCAGCCATGTACGTTGCTATCCAGGCTGTGCTATCCCTGTACGCCTCTGGCCGTACCACTGGCATCGTGATGGACTCCGGTGACGGGGTCACCCACACTGTGCCCATCTACGAGGGGTATGCCCTCCCCCATGCCATCCTGCGTCTGGACCTGGCTGGCCGGGACCTGACTGACTACCTCATGAAGATCCTCACCGAGCGCGGCTACAGCTTCACCACCACGGCCGAGCGGGAAATCGTGCGTGACATTAAGGAGAAGCTGTGCTACGTCGCCCTGGACTTCGAGCAAGAGATGGCCACGGCTGCTTCCAGCTCCTCCCTGGAGAAGAGCTACGAGCTGCCTGACGGCCAGGTCATCACCATTGGCAATGAGCGGTTCCGCTGCCCTGAGGCACTCTTCCAGCCTTCCTTCCTGGGTGAGTGGAGACTGTCTCCCGGCTCTGCCTGACATGAGGGTTACCCCTCGGGGCTGTGCTGTGGAAGCTAAGTCCTGCCCTCATTTCCCTCTCAGGCATGGAGTCCTGTGGCATCCACGAAACTACCTTCAACTCCATCATGAAGTGTGACGTGGACATCCGCAAAGACCTGTACGCCAACACAGTGCTGTCTGGCGGCACCACCATGTACCCTGGCATTGCCGACAGGATGCAGAAGGAGATCACTGCCCTGGCACCCAGCACAATGAAGATCAAGGTGGGTGTCTTTCCTGCCTGAGCTGACCTGGGCAGGTCGGCTGTGGGGTCCTGTGGTGTGTGGGGAGCTGTCACATCCAGGGTCCTCACTGCCTGTCCCCTTCCCTCCTCAGATCATTGCTCCTCCTGAGCGCAAGTACTCCGTGTGGATCGGCGGCTCCATCCTGGCCTCGCTGTCCACCTTCCAGCAGATGTGGATCAGCAAGCAGGAGTATGACGAGTCCGGCCCCTCCATCGTCCACCGCAAATGCTTCTAGGCGGACTATGACTTAGTTGCGTTACACCCTTTCTTGACAAAACCTAACTTGCGCAGAAAACAAGATGAGATTGGCATGGCTTTATTTGTTTTTTTTGTTTTGTTTTGGTTTTTTTTTTTTTTTTGGCTTGACTCAGGATTTAAAAACTGGAACGGTGAAGGTGACAGCAGTCGGTTGGAGCGAGCATCCCCCAAAGTTCACAATGTGGCCGAGGACTTTGATTGCACATTGTTGTTTTTTTAATAGTCATTCCAAATATGAGATGCGTTGTTACAGGAAGTCCCTTGCCATCCTAAAAGCCACCCCACTTCTCTCTAAGGAGAATGGCCCAGTCCTCTCCCAAGTCCACACAGGGGAGGTGATAGCATTGCTTTCGTGTAAATTATGTAATGCAAAATTTTTTTAATCTTCGCCTTAATACTTTTTTATTTTGTTTTATTTTGAATGATGAGCCTTCGTGCCCCCCCTTCCCCCTTTTTTGTCCCCCAACTTGAGATGTATGAAGGCTTTTGGTCTCCCTGGGAGTGGGTGGAGGCAGCCAGGGCTTACCTGTACACTGACTTGAGACCAGTTGAATAAAAGTGCACACCTTAAAAATGAGGCCAAG\"\n",
        "elif seq == \"GAPDH pre-mRNA\":\n",
        "  seq = \"GCTCTCTGCTCCTCCTGTTCGACAGTCAGCCGCATCTTCTTTTGCGTCGCCAGGTGAAGACGGGCGGAGAGAAACCCGGGAGGCTAGGGACGGCCTGAAGGCGGCAGGGGCGGGCGCAGGCCGGATGTGTTCGCGCCGCTGCGGGGTGGGCCCGGGCGGCCTCCGCATTGCAGGGGCGGGCGGAGGACGTGATGCGGCGCGGGCTGGGCATGGAGGCCTGGTGGGGGAGGGGAGGGGAGGCGTGTGTGTCGGCCGGGGCCACTAGGCGCTCACTGTTCTCTCCCTCCGCGCAGCCGAGCCACATCGCTCAGACACCATGGGGAAGGTGAAGGTCGGAGTCAACGGGTGAGTTCGCGGGTGGCTGGGGGGCCCTGGGCTGCGACCGCCCCCGAACCGCGTCTACGAGCCTTGCGGGCTCCGGGTCTTTGCAGTCGTATGGGGGCAGGGTAGCTGTTCCCCGCAAGGAGAGCTCAAGGTCAGCGCTCGGACCTGGCGGAGCCCCGCACCCAGGCTGTGGCGCCCTGTGCAGCTCCGCCCTTGCGGCGCCATCTGCCCGGAGCCTCCTTCCCCTAGTCCCCAGAAACAGGAGGTCCCTACTCCCGCCCGAGATCCCGACCCGGACCCCTAGGTGGGGGACGCTTTCTTTCCTTTCGCGCTCTGCGGGGTCACGTGTCGCAGAGGAGCCCCTCCCCCACGGCCTCCGGCACCGCAGGCCCCGGGATGCTAGTGCGCAGCGGGTGCATCCCTGTCCGGATGCTGCGCCTGCGGTAGAGCGGCCGCCATGTTGCAACCGGGAAGGAAATGAATGGGCAGCCGTTAGGAAAGCCTGCCGGTGACTAACCCTGCGCTCCTGCCTCGATGGGTGGAGTCGCGTGTGGCGGGGAAGTCAGGTGGAGCGAGGCTAGCTGGCCCGATTTCTCCTCCGGGTGATGCTTTTCCTAGATTATTCTCTGGTAAATCAAAGAAGTGGGTTTATGGAGGTCCTCTTGTGTCCCCTCCCCGCAGAGGTGTGGTGGCTGTGGCATGGTGCCAAGCCGGGAGAAGCTGAGTCATGGGTAGTTGGAAAAGGACATTTCCACCGCAAAATGGCCCCTCTGGTGGTGGCCCCTTCCTGCAGCGCCGGCTCACCTCACGGCCCCGCCCTTCCCCTGCCAGCCTAGCGTTGACCCGACCCCAAAGGCCAGGCTGTAAATGTCACCGGGAGGATTGGGTGTCTGGGCGCCTCGGGGAACCTGCCCTTCTCCCCATTCCGTCTTCCGGAAACCAGATCTCCCACCGCACCCTGGTCTGAGGTTAAATATAGCTGCTGACCTTTCTGTAGCTGGGGGCCTGGGCTGGGGCTCTCTCCCATCCCTTCTCCCCACACACATGCACTTACCTGTGCTCCCACTCCTGATTTCTGGAAAAGAGCTAGGAAGGACAGGCAACTTGGCAAATCAAAGCCCTGGGACTAGGGGGTTAAAATACAGCTTCCCCTCTTCCCACCCGCCCCAGTCTCTGTCCCTTTTGTAGGAGGGACTTAGAGAAGGGGTGGGCTTGCCCTGTCCAGTTAATTTCTGACCTTTACTCCTGCCCTTTGAGTTTGATGATGCTGAGTGTACAAGCGTTTTCTCCCTAAAGGGTGCAGCTGAGCTAGGCAGCAGCAAGCATTCCTGGGGTGGCATAGTGGGGTGGTGAATACCATGTACAAAGCTTGTGCCCAGACTGTGGGTGGCAGTGCCCCACATGGCCGCTTCTCCTGGAAGGGCTTCGTATGACTGGGGGTGTTGGGCAGCCCTGGAGCCTTCAGTTGCAGCCATGCCTTAAGCCAGGCCAGCCTGGCAGGGAAGCTCAAGGGAGATAAAATTCAACCTCTTGGGCCCTCCTGGGGGTAAGGAGATGCTGCATTCGCCCTCTTAATGGGGAGGTGGCCTAGGGCTGCTCACATATTCTGGAGGAGCCTCCCCTCCTCATGCCTTCTTGCCTCTTGTCTCTTAGATTTGGTCGTATTGGGCGCCTGGTCACCAGGGCTGCTTTTAACTCTGGTAAAGTGGATATTGTTGCCATCAATGACCCCTTCATTGACCTCAACTACATGGTGAGTGCTACATGGTGAGCCCCAAAGCTGGTGTGGGAGGAGCCACCTGGCTGATGGGCAGCCCCTTCATACCCTCACGTATTCCCCCAGGTTTACATGTTCCAATATGATTCCACCCATGGCAAATTCCATGGCACCGTCAAGGCTGAGAACGGGAAGCTTGTCATCAATGGAAATCCCATCACCATCTTCCAGGAGTGAGTGGAAGACAGAATGGAAGAAATGTGCTTTGGGGAGGCAACTAGGATGGTGTGGCTCCCTTGGGTATATGGTAACCTTGTGTCCCTCAATATGGTCCTGTCCCCATCTCCCCCCCACCCCCATAGGCGAGATCCCTCCAAAATCAAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACTGGCGTCTTCACCACCATGGAGAAGGCTGGGGTGAGTGCAGGAGGGCCCGCGGGAGGGGAAGCTGACTCAGCCCTGCAAAGGCAGGACCCGGGTTCATAACTGTCTGCTTCTCTGCTGTAGGCTCATTTGCAGGGGGGAGCCAAAAGGGTCATCATCTCTGCCCCCTCTGCTGATGCCCCCATGTTCGTCATGGGTGTGAACCATGAGAAGTATGACAACAGCCTCAAGATCATCAGGTGAGGAAGGCAGGGCCCGTGGAGAAGCGGCCAGCCTGGCACCCTATGGACACGCTCCCCTGACTTGCGCCCCGCTCCCTCTTTCTTTGCAGCAATGCCTCCTGCACCACCAACTGCTTAGCACCCCTGGCCAAGGTCATCCATGACAACTTTGGTATCGTGGAAGGACTCATGGTATGAGAGCTGGGGAATGGGACTGAGGCTCCCACCTTTCTCATCCAAGACTGGCTCCTCCCTGCCGGGGCTGCGTGCAACCCTGGGGTTGGGGGTTCTGGGGACTGGCTTTCCCATAATTTCCTTTCAAGGTGGGGAGGGAGGTAGAGGGGTGATGTGGGGAGTACGCTGCAGGGCCTCACTCCTTTTGCAGACCACAGTCCATGCCATCACTGCCACCCAGAAGACTGTGGATGGCCCCTCCGGGAAACTGTGGCGTGATGGCCGCGGGGCTCTCCAGAACATCATCCCTGCCTCTACTGGCGCTGCCAAGGCTGTGGGCAAGGTCATCCCTGAGCTGAACGGGAAGCTCACTGGCATGGCCTTCCGTGTCCCCACTGCCAACGTGTCAGTGGTGGACCTGACCTGCCGTCTAGAAAAACCTGCCAAATATGATGACATCAAGAAGGTGGTGAAGCAGGCGTCGGAGGGCCCCCTCAAGGGCATCCTGGGCTACACTGAGCACCAGGTGGTCTCCTCTGACTTCAACAGCGACACCCACTCCTCCACCTTTGACGCTGGGGCTGGCATTGCCCTCAACGACCACTTTGTCAAGCTCATTTCCTGGTATGTGGCTGGGGCCAGAGACTGGCTCTTAAAAAGTGCAGGGTCTGGCGCCCTCTGGTGGCTGGCTCAGAAAAAGGGCCCTGACAACTCTTTTCATCTTCTAGGTATGACAACGAATTTGGCTACAGCAACAGGGTGGTGGACCTCATGGCCCACATGGCCTCCAAGGAGTAAGACCCCTGGACCACCAGCCCCAGCAAGAGCACAAGAGGAAGAGAGAGACCCTCACTGCTGGGGAGTCCCTGCCACACTCAGTCCCCCACCACACTGAATCTCCCCTCCTCACAGTTGCCATGTAGACCCCTTGAAGAGGGGAGGGGCCTAGGGAGCCGCACCTTGTCATGTACCATCAATAAAGTACCCTGTGCTCAACCA\"\n",
        "elif seq == \"EGFP + CENPB 3' UTR\":\n",
        "  seq = \"ATGGTGAGCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGTAAGTCACTGGACCTAGCTGTGCCCCCAACCTAGATTGGCAGCACCACCCCAGGGCAGAGGACTCTCTGGGCACCCGCTGTGCATGGAGCCAGAGTGCAGAGCCCCAGATCCTTTAGTAATGCTTCCCCTGGTCCTGCAACAGGCCCGGTCACCTCGGCCGGGCCCGGGGCTGAGGTCAGCCTCACTGCCTGCTTATTGCCTCTTTCTCAGAATCCTCTTTCCTCCCCATTTGGCCCTGGGCTCAGGGGACCAGGTGGGGCGGGTGGGGAGCTGTCCGGTGCTACCACACCGTGCCCTCAGTGGACTAACCACAGCAGCAGCCAGGGATGGGCCCTGGAGGTTCCCGGCCGGAGAGTGCCTCTCCCCTCTGCCATCCACGTCAGGTCTTTGGTGGGGGGACCCCAAAGCCATTCTGGGAAGGGCTCCAGAAGAAGGTCCAGCCTAGGCCCCCTGCAAGGCTGGCAGCCCCCACCCCCACCCCCCAGGCCGCCTTGAGAAGCACAGTTTAACTCACTGCGGGCTCCTGAGCCTGCTTCTGCCTGCTTTCCACCTCCCCAGTCCCTTTCTCTGGCCCTGTCCATGTGACTTTGGCCCTTGGTTTTCTTTCCAGATTGGAGGTTTCCAAGAGGCCCCCCACCGTGGAAGTAACCAAGGGCGCTTCCTTGTGGGCAGCTGCAGGCCCCATGCCTCTCCTCCCTCTCTGGCAGGGCCCCATCCTGGGCAGAGGGGCCTGGGGCTGGGCCCAGAGTCCAGCCGTCCAGCTGCTCCTTTCCCAGTTTGATTTCAATAAATCTGTCCACTCCCCTTTTGTGGGGGTGAACGTTTTAACAGCCAAGGGT\"\n",
        "elif seq == \"Enter your sequence or select a sample sequence (default: ACTB pre-mRNA)\":\n",
        "  seq = \"CGAGACCGCGTCCGCCCCGCGAGCACAGAGCCTCGCCTTTGCCGATCCGCCGCCCGTCCACACCCGCCGCCAGGTAAGCCCGGCCAGCCGACCGGGGCAGGCGGCTCACGGCCCGGCCGCAGGCGGCCGCGGCCCCTTCGCCCGTGCAGAGCCGCCGTCTGGGCCGCAGCGGGGGGCGCATGGGGGGGGAACCGGACCGCCGTGGGGGGCGCGGGAGAAGCCCCTGGGCCTCCGGAGATGGGGGACACCCCACGCCAGTTCGGAGGCGCGAGGCCGCGCTCGGGAGGCGCGCTCCGGGGGTGCCGCTCTCGGGGCGGGGGCAACCGGCGGGGTCTTTGTCTGAGCCGGGCTCTTGCCAATGGGGATCGCAGGGTGGGCGCGGCGGAGCCCCCGCCAGGCCCGGTGGGGGCTGGGGCGCCATTGCGCGTGCGCGCTGGTCCTTTGGGCGCTAACTGCGTGCGCGCTGGGAATTGGCGCTAATTGCGCGTGCGCGCTGGGACTCAAGGCGCTAACTGCGCGTGCGTTCTGGGGCCCGGGGTGCCGCGGCCTGGGCTGGGGCGAAGGCGGGCTCGGCCGGAAGGGGTGGGGTCGCCGCGGCTCCCGGGCGCTTGCGCGCACTTCCTGCCCGAGCCGCTGGCCGCCCGAGGGTGTGGCCGCTGCGTGCGCGCGCGCCGACCCGGCGCTGTTTGAACCGGGCGGAGGCGGGGCTGGCGCCCGGTTGGGAGGGGGTTGGGGCCTGGCTTCCTGCCGCGCGCCGCGGGGACGCCTCCGACCAGTGTTTGCCTTTTATGGTAATAACGCGGCCGGCCCGGCTTCCTTTGTCCCCAATCTGGGCGCGCGCCGGCGCCCCCTGGCGGCCTAAGGACTCGGCGCGCCGGAAGTGGCCAGGGCGGGGGCGACCTCGGCTCACAGCGCGCCCGGCTATTCTCGCAGCTCACCATGGATGATGATATCGCCGCGCTCGTCGTCGACAACGGCTCCGGCATGTGCAAGGCCGGCTTCGCGGGCGACGATGCCCCCCGGGCCGTCTTCCCCTCCATCGTGGGGCGCCCCAGGCACCAGGTAGGGGAGCTGGCTGGGTGGGGCAGCCCCGGGAGCGGGCGGGAGGCAAGGGCGCTTTCTCTGCACAGGAGCCTCCCGGTTTCCGGGGTGGGGGCTGCGCCCGTGCTCAGGGCTTCTTGTCCTTTCCTTCCCAGGGCGTGATGGTGGGCATGGGTCAGAAGGATTCCTATGTGGGCGACGAGGCCCAGAGCAAGAGAGGCATCCTCACCCTGAAGTACCCCATCGAGCACGGCATCGTCACCAACTGGGACGACATGGAGAAAATCTGGCACCACACCTTCTACAATGAGCTGCGTGTGGCTCCCGAGGAGCACCCCGTGCTGCTGACCGAGGCCCCCCTGAACCCCAAGGCCAACCGCGAGAAGATGACCCAGGTGAGTGGCCCGCTACCTCTTCTGGTGGCCGCCTCCCTCCTTCCTGGCCTCCCGGAGCTGCGCCCTTTCTCACTGGTTCTCTCTTCTGCCGTTTTCCGTAGGACTCTCTTCTCTGACCTGAGTCTCCTTTGGAACTCTGCAGGTTCTATTTGCTTTTTCCCAGATGAGCTCTTTTTCTGGTGTTTGTCTCTCTGACTAGGTGTCTAAGACAGTGTTGTGGGTGTAGGTACTAACACTGGCTCGTGTGACAAGGCCATGAGGCTGGTGTAAAGCGGCCTTGGAGTGTGTATTAAGTAGGTGCACAGTAGGTCTGAACAGACTCCCCATCCCAAGACCCCAGCACACTTAGCCGTGTTCTTTGCACTTTCTGCATGTCCCCCGTCTGGCCTGGCTGTCCCCAGTGGCTTCCCCAGTGTGACATGGTGTATCTCTGCCTTACAGATCATGTTTGAGACCTTCAACACCCCAGCCATGTACGTTGCTATCCAGGCTGTGCTATCCCTGTACGCCTCTGGCCGTACCACTGGCATCGTGATGGACTCCGGTGACGGGGTCACCCACACTGTGCCCATCTACGAGGGGTATGCCCTCCCCCATGCCATCCTGCGTCTGGACCTGGCTGGCCGGGACCTGACTGACTACCTCATGAAGATCCTCACCGAGCGCGGCTACAGCTTCACCACCACGGCCGAGCGGGAAATCGTGCGTGACATTAAGGAGAAGCTGTGCTACGTCGCCCTGGACTTCGAGCAAGAGATGGCCACGGCTGCTTCCAGCTCCTCCCTGGAGAAGAGCTACGAGCTGCCTGACGGCCAGGTCATCACCATTGGCAATGAGCGGTTCCGCTGCCCTGAGGCACTCTTCCAGCCTTCCTTCCTGGGTGAGTGGAGACTGTCTCCCGGCTCTGCCTGACATGAGGGTTACCCCTCGGGGCTGTGCTGTGGAAGCTAAGTCCTGCCCTCATTTCCCTCTCAGGCATGGAGTCCTGTGGCATCCACGAAACTACCTTCAACTCCATCATGAAGTGTGACGTGGACATCCGCAAAGACCTGTACGCCAACACAGTGCTGTCTGGCGGCACCACCATGTACCCTGGCATTGCCGACAGGATGCAGAAGGAGATCACTGCCCTGGCACCCAGCACAATGAAGATCAAGGTGGGTGTCTTTCCTGCCTGAGCTGACCTGGGCAGGTCGGCTGTGGGGTCCTGTGGTGTGTGGGGAGCTGTCACATCCAGGGTCCTCACTGCCTGTCCCCTTCCCTCCTCAGATCATTGCTCCTCCTGAGCGCAAGTACTCCGTGTGGATCGGCGGCTCCATCCTGGCCTCGCTGTCCACCTTCCAGCAGATGTGGATCAGCAAGCAGGAGTATGACGAGTCCGGCCCCTCCATCGTCCACCGCAAATGCTTCTAGGCGGACTATGACTTAGTTGCGTTACACCCTTTCTTGACAAAACCTAACTTGCGCAGAAAACAAGATGAGATTGGCATGGCTTTATTTGTTTTTTTTGTTTTGTTTTGGTTTTTTTTTTTTTTTTGGCTTGACTCAGGATTTAAAAACTGGAACGGTGAAGGTGACAGCAGTCGGTTGGAGCGAGCATCCCCCAAAGTTCACAATGTGGCCGAGGACTTTGATTGCACATTGTTGTTTTTTTAATAGTCATTCCAAATATGAGATGCGTTGTTACAGGAAGTCCCTTGCCATCCTAAAAGCCACCCCACTTCTCTCTAAGGAGAATGGCCCAGTCCTCTCCCAAGTCCACACAGGGGAGGTGATAGCATTGCTTTCGTGTAAATTATGTAATGCAAAATTTTTTTAATCTTCGCCTTAATACTTTTTTATTTTGTTTTATTTTGAATGATGAGCCTTCGTGCCCCCCCTTCCCCCTTTTTTGTCCCCCAACTTGAGATGTATGAAGGCTTTTGGTCTCCCTGGGAGTGGGTGGAGGCAGCCAGGGCTTACCTGTACACTGACTTGAGACCAGTTGAATAAAAGTGCACACCTTAAAAATGAGGCCAAG\"\n",
        "else: # user input \n",
        "  # clean up sequence: upper case, remove space, RNA->DNA\n",
        "  seq = seq.upper().replace('U','T').replace(' ','')\n",
        "  # if contains non ACGT letters:\n",
        "  if not all(char in 'ACGT' for char in seq):\n",
        "    print(\"\\n\\n!!!ERROR: Your sequence contains non A/C/G/T letters!\")\n",
        "    exit()\n",
        "\n",
        "######### options\n",
        "\n",
        "# set model type\n",
        "MODEL_TYPE = \"Usage\" #@param [\"Usage\", \"P(splice)\"]\n",
        "#@markdown - *Usage*: predict tissue-specific usage of a splice site (default)\n",
        "#@markdown - *P(splice)*: predict tissue-specific strength of a splice site\n",
        "\n",
        "if MODEL_TYPE == \"Usage\":\n",
        "  model_nums = [1,3,5,7]\n",
        "else:\n",
        "  model_nums = [0,2,4,6]\n",
        "# 0 = Heart, P(splice)\n",
        "# 1 = Heart, usage\n",
        "# 2 = Liver, P(splice)\n",
        "# 3 = Liver, usage\n",
        "# 4 = Brain, P(splice)\n",
        "# 5 = Brain, usage\n",
        "# 6 = Testis, P(splice)\n",
        "# 7 = Testis, usage\n",
        "\n",
        "tissues = ['Heart','Liver','Brain','Testis']\n",
        "\n",
        "\n",
        "# min score\n",
        "MIN_SCORE = 0.1 #@param {type:\"number\"}\n",
        "#@markdown - Minimum usage/strength of a site to be included in the output \n",
        "\n",
        "if not os.path.exists(\"splice-site-prediction-with-pangolin\"):\n",
        "  print(\"\\n\\n\")\n",
        "  print(datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\"),': Loading Pangolin (~50s, first time only)')\n",
        "  print(\"~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\")\n",
        "  #!pip install pyvcf gffutils biopython pandas pyfastx\n",
        "  !git clone https://github.com/xuebingwu/splice-site-prediction-with-pangolin.git\n",
        "  %cd splice-site-prediction-with-pangolin\n",
        "  !pip install .\n",
        "  %cd /content\n",
        "\n",
        "from pkg_resources import resource_filename\n",
        "from pangolin.model import *\n",
        "\n",
        "# Load models\n",
        "def load_models(model_nums):\n",
        "    models = []\n",
        "    for i in model_nums:\n",
        "        for j in range(1, 6):\n",
        "            model = Pangolin(L, W, AR)\n",
        "            if torch.cuda.is_available():\n",
        "                model.cuda()\n",
        "                weights = torch.load(resource_filename(\"pangolin\",\"models/final.%s.%s.3\" % (j, i)))\n",
        "            else:\n",
        "                weights = torch.load(resource_filename(\"pangolin\",\"models/final.%s.%s.3\" % (j, i)),\n",
        "                                     map_location=torch.device('cpu'))\n",
        "            model.load_state_dict(weights)\n",
        "            model.eval()\n",
        "            models.append(model)\n",
        "    return models\n",
        "\n",
        "IN_MAP = np.asarray([[0, 0, 0, 0],\n",
        "                     [1, 0, 0, 0],\n",
        "                     [0, 1, 0, 0],\n",
        "                     [0, 0, 1, 0],\n",
        "                     [0, 0, 0, 1]])\n",
        "INDEX_MAP = {0:1, 1:2, 2:4, 3:5, 4:7, 5:8, 6:10, 7:11}\n",
        "\n",
        "def one_hot_encode(seq, strand):\n",
        "    '''\n",
        "    one hot enencode of a dna sequence\n",
        "    strand: + or -. - for reverse complement\n",
        "    '''\n",
        "    # replace space\n",
        "    seq = seq.replace(' ','')\n",
        "    seq = seq.upper().replace('A', '1').replace('C', '2')\n",
        "    seq = seq.replace('G', '3').replace('T', '4').replace('N', '0')\n",
        "    if strand == '+':\n",
        "        seq = np.asarray(list(map(int, list(seq))))\n",
        "    elif strand == '-':\n",
        "        seq = np.asarray(list(map(int, list(seq[::-1]))))\n",
        "        seq = (5 - seq) % 5  # Reverse complement\n",
        "    return IN_MAP[seq.astype('int8')]\n",
        "\n",
        "def run_models(seq,strand):\n",
        "    '''\n",
        "    run Pangolin models on an input sequence seq\n",
        "    \n",
        "    input:\n",
        "      seq     - a sequence\n",
        "      strand  - reverse complement seq if set to '-'\n",
        "    \n",
        "    output:\n",
        "      scores  - a list of arrays, each array is a tissue,\n",
        "              - each element in the array is the score for a nucleotide\n",
        "    '''\n",
        "\n",
        "    seq = one_hot_encode(seq,strand).T\n",
        "    seq = torch.from_numpy(np.expand_dims(seq, axis=0)).float()\n",
        "\n",
        "    if torch.cuda.is_available():\n",
        "        seq = seq.to(torch.device(\"cuda\"))\n",
        "\n",
        "    scores = []\n",
        "    for j, model_num in enumerate(model_nums):\n",
        "        score = []\n",
        "        # Average across 5 models\n",
        "        for model in models[5*j:5*j+5]:\n",
        "            with torch.no_grad():\n",
        "                score.append(model(seq)[0][INDEX_MAP[model_num],:].cpu().numpy())\n",
        "        scores.append(np.mean(score, axis=0))\n",
        "    return scores\n",
        "\n",
        "def filter_predicted_sites(seq,scores,min_score):\n",
        "    # filter sites based on max score across 4 tissues\n",
        "    # annotate whether a site is 5'SS or 3'SS\n",
        "    # note that Pangolin does not train 5'SS/3'SS separately\n",
        "\n",
        "    # output: sites\n",
        "    #   - a dictionary from position to site info\n",
        "    #   - e.g. \n",
        "    #     53:\"TGCGTCGCCAGgtgaagacg\t5'SS\t0.803,0.78,0.826,0.712\"\n",
        "\n",
        "    sites = {}\n",
        "    for i in range(len(seq)):\n",
        "        # filter based on max score in 4 tissues\n",
        "        if scores[0][i] >= min_score or scores[1][i] >= min_score or scores[2][i] >= min_score or scores[3][i] >= min_score:\n",
        "            sites[i+1] = seq[i-10:i+10]\n",
        "            # determine splice site type\n",
        "            if seq[i+1:i+3] == 'GT' and seq[i-2:i] != 'AG':\n",
        "                # 5'SS: GT but not AG\n",
        "                sites[i+1] = sites[i+1][:11].upper()+sites[i+1][11:].lower()\n",
        "                sites[i+1] = sites[i+1] + \"\\t5'SS\"\n",
        "            elif seq[i+1:i+3] != 'GT' and seq[i-2:i] == 'AG':\n",
        "                # 3'SS: AG but not GT\n",
        "                sites[i+1] = sites[i+1][:10].lower()+sites[i+1][10:].upper()\n",
        "                sites[i+1] = sites[i+1] + \"\\t3'SS\"\n",
        "            elif seq[i+1:i+3] == 'GT' and seq[i-2:i] == 'AG':\n",
        "                # very rarely, has both GT and AG\n",
        "                sites[i+1] = sites[i+1] + '\\tUndetermined'\n",
        "            else:\n",
        "                # very rarely, no GT or AG\n",
        "                sites[i+1] = sites[i+1] + '\\tNoncanonical'\n",
        "            sites[i+1] = sites[i+1] + '\\t'+ ','.join([str(round(scores[0][i],3)),str(round(scores[1][i],3)),str(round(scores[2][i],3)),str(round(scores[3][i],3))]) \n",
        "    return sites\n",
        "\n",
        "\n",
        "# plot sites\n",
        "def plot_sites(sites,L):\n",
        "  # sites - filtered splice sites\n",
        "  # L: length of the input sequence\n",
        "\n",
        "  # setup the plot\n",
        "  pylab.plot()\n",
        "  pylab.xlim(0,L)\n",
        "  pylab.ylim(-0.05,1.2)\n",
        "  pylab.ylabel(MODEL_TYPE)\n",
        "  pylab.xlabel('Position')\n",
        "\n",
        "  # legend/text for tissue\n",
        "  colors = ['red','blue','green','orange']\n",
        "  label_xpos = [L/40,L/7,1.7*L/7,2.4*L/7] # position of the tissue label\n",
        "  for i in range(4):\n",
        "    pylab.text(label_xpos[i],1.1, tissues[i], color=colors[i])\n",
        "\n",
        "  # legend for site type\n",
        "  markers = {\"5'SS\":\">\",\"3'SS\":'<',\"Undetermined\":\"D\",\"Noncanonical\":\"s\"}\n",
        "  legend_handles = []\n",
        "  for site_type in [\"5'SS\",\"3'SS\"]: #,\"Undetermined\",\"Noncanonical\"]:\n",
        "    legend_handles.append(mlines.Line2D([], [], color='black', marker=markers[site_type], linestyle='None',\n",
        "                           label=site_type))\n",
        "  pylab.legend(handles=legend_handles,ncol=4)\n",
        "\n",
        "  # a thick line at the bottom to represent the entire input sequence\n",
        "  pylab.plot([0,L],[-0.02,-0.02],linewidth=7,c='black')\n",
        "\n",
        "  # plot each site\n",
        "  for pos in sites:\n",
        "    x = int(pos)\n",
        "    site_seq, site_type, site_score = sites[pos].split('\\t')\n",
        "    site_scores = site_score.split(',')\n",
        "    for i in range(4):\n",
        "      # a marker\n",
        "      pylab.plot(x,float(site_scores[i]),markers[site_type],c=colors[i])\n",
        "      # a vertical line\n",
        "      pylab.plot([x,x],[0,float(site_scores[i])],linewidth=0.5,c='gray')\n",
        "\n",
        "  # save as pdf\n",
        "  pylab.savefig('predicted_splice_sites.pdf')  \n",
        "  pylab.show()\n",
        "\n",
        "## start the analysis\n",
        "\n",
        "print(\"\\n\")\n",
        "print(datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\"),': Loading models')\n",
        "#load models\n",
        "models = load_models(model_nums)\n",
        "\n",
        "print(datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\"),': Scoring sites')\n",
        "# score sequences\n",
        "scores = run_models( 5000*'N'+ seq + 5000*'N','+')\n",
        "\n",
        "print(datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\"),': Filtering sites')\n",
        "sites = filter_predicted_sites(seq,scores,MIN_SCORE)\n",
        "\n",
        "# plot\n",
        "print(datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\"),': Plotting\\n')\n",
        "plot_sites(sites,len(seq))\n",
        "\n",
        "# download figure\n",
        "files.download(f\"predicted_splice_sites.pdf\")\n",
        "\n",
        "# text/table output\n",
        "print('\\n\\nPredicted splice sites\\n')\n",
        "print('\\t'.join(['Position','Sequence','Type'])+'\\t'+'\\t'.join(tissues))\n",
        "print('-----------------------------------------------------------------------')\n",
        "for pos in sites:\n",
        "  print(pos,'\\t',sites[pos].replace(',','\\t'))\n",
        "\n",
        "print(\"\\nColumns:\")\n",
        "print(\"Position:\\tSite position in the sequence\")\n",
        "print(\"Sequence:\\t20-nt junction sequence. Intron-lower case; Exon-upper case\")\n",
        "print(\"Type    :\\t5' splice site (5'SS), 3' splice site (3'SS), or Undetermined, or Noncanonical\")\n",
        "print(\"Scores  :\\tSite usage or P(splice) in each tissue\")\n"
      ],
      "metadata": {
        "id": "D6znA5iqZu1b",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}